File: Leetcode/230/230.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int counter, k, answer;
    bool stopRecursion;
    void f(TreeNode* root) {
        if (!stopRecursion && root) {
            f(root->left);
            ++counter;
            if (counter == k) {
                answer = root->val;
                stopRecursion = true;
                return;
            }
            f(root->right);
        }
    }
    int kthSmallest(TreeNode* root, int k) {
        this->k = k;
        this->counter = 0;
        this->stopRecursion = false;
        f(root);
        return this->answer;
    }
};

File: Leetcode/230/230.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func dfs(root *TreeNode, k *int, haveAns *bool, answer *int) {
	if *haveAns {
		return
	}
	if root != nil {
		dfs(root.Left, k, haveAns, answer)
        (*k)--
		if *k == 0 {
			*answer = root.Val
			*haveAns = true
			return
		}
		dfs(root.Right, k, haveAns, answer)
	}
}

func kthSmallest(root *TreeNode, k int) int {
    haveAns := false
    var answer int
	dfs(root, &k, &haveAns, &answer)
	return answer
}